src/MongoDB.EntityFrameworkCore/Storage/MongoClientWrapper.cs

```diff
@@ -99,138 +101,278 @@ public async Task<IClientSessionHandle> StartSessionAsync(CancellationToken canc
+        var searchIndexModelsMap = new Dictionary<string, List<CreateSearchIndexModel>?>();
+
         var existed = await DatabaseExistsAsync(cancellationToken).ConfigureAwait(false);
-        using var collectionNamesCursor =
-            await Database.ListCollectionNamesAsync(null, cancellationToken).ConfigureAwait(false);
-        var existingCollectionNames = collectionNamesCursor.ToList(cancellationToken);
+        using var collectionNamesCursor = await Database.ListCollectionNamesAsync(null, cancellationToken).ConfigureAwait(false);
+        var collectionNames = await collectionNamesCursor.ToListAsync(cancellationToken);
 
         foreach (var entityType in model.Model.GetEntityTypes().Where(e => e.IsDocumentRoot()))
         {
             var collectionName = entityType.GetCollectionName();
-
-            if (!existingCollectionNames.Contains(collectionName))
+            if (!collectionNames.Contains(collectionName))
             {
+                collectionNames.Add(collectionName);
                 try
                 {
-                    await Database.CreateCollectionAsync(collectionName, cancellationToken: cancellationToken)
-                        .ConfigureAwait(false);
-                    existingCollectionNames.Add(collectionName);
+                    await Database.CreateCollectionAsync(collectionName, null, cancellationToken).ConfigureAwait(false);
                 }
                 catch (MongoCommandException ex) when (ex.Message.Contains("already exists"))
                 {
+                    // Ignore collection already exists in cases of concurrent creation
                 }
             }
 
-            var indexManager = Database.GetCollection<BsonDocument>(entityType.GetCollectionName()).Indexes;
-            var indexCursor = await indexManager.ListAsync(cancellationToken).ConfigureAwait(false);
-            var existingIndexNames = indexCursor.ToList(cancellationToken).Select(i => i["name"].AsString).ToList();
-            await CreateIndexesAsync(entityType, indexManager, existingIndexNames, [], cancellationToken).ConfigureAwait(false);
+            if (!existingIndexesMap.TryGetValue(collectionName, out var indexes))
+            {
+                using var cursor = await Database.GetCollection<BsonDocument>(collectionName)
+                    .Indexes.ListAsync(cancellationToken).ConfigureAwait(false);
+
+                indexes = (await cursor.ToListAsync(cancellationToken)).Select(i => i["name"].AsString).ToList();
+                existingIndexesMap[collectionName] =  indexes;
+            }
+
+            if (!existingSearchIndexesMap.TryGetValue(collectionName, out var searchIndexes))
+            {
+                using var cursor = await Database.GetCollection<BsonDocument>(collectionName)
+                    .SearchIndexes.ListAsync(cancellationToken: cancellationToken).ConfigureAwait(false);
```

Comment:
I realized on seeing tests fail in Evergreen that we need to avoid making calls to Atlas-only features when the application is not using Atlas. Right now, this causes EnsureCreated to throw against any non-Atlas server. Instead, I hope to change this so that it asks the server for vector indexes _only if the app has defined some vector indexes in its model_.